Flow ebos fix fip

andlaus opened the pull request:
this should fix #988, but please test before merge.

andlaus commented:
jenkins run norne_ebos please

blattms commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -992,29 +993,41 @@ namespace Opm {
                 const auto& fs = intQuants.fluidState();
+                const double pv =
+                    ebosSimulator_.model().dofTotalVolume(cellIdx)
+                    * ebosSimulator_.problem().porosity(cellIdx);
```
These are a lot of new lines of code. As far as I can judge the only change is that you now calculate the pore volume from ebos properties instead of taking the one stored in the properties. Is this the fix, i.e. are the values in `geo_` wrong. If not, what is the reason to obfuscate it with DUNE/ebos lingo and why is it even part of this PR?

blattms commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1056,25 +1069,31 @@ namespace Opm {
+                        pres[region] += pv * fs.pressure(FluidSystem::oilPhaseIdx).value();
                     }
                 }
                 for (int c = 0; c < nc; ++c) {
```
Here you still use the old loop style and the obfuscated one from above. Why? Seems inconsistent and will raise questions in the future.

blattms replied:
BTW and totally unrelated. It seem we have still not fixed all region summations in parallel runs (e.g. by masking or as you did above by only computing on InteriorNodes. But that fix should be in a separate PR.

andlaus replied:
> Here you still use the old loop style and the obfuscated one from above. Why?

obfuscation lies in the eye of the beholder ;). why? because it means less lines of changed code (didn't somebody complain about this recently) and at this point, one can assume that the cache for intensive quantities is "hot". but you're probably right that it should be changed to a loop over the grid because as-is it does not work with an overlap. (the same is the case for flow_legacy as far as I can see.)

andlaus replied:
> BTW and totally unrelated. It seem we have still not fixed all region summations in parallel runs (e.g. by masking or as you did above by only computing on InteriorNodes. But that fix should be in a separate PR.

that's quite possible. the (sequential) FIP numbers after this PR are consistent, with flow_legacy, though...

blattms commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1136,15 +1158,18 @@ namespace Opm {
+                        const double pv =
+                            ebosSimulator_.model().dofTotalVolume(c)
+                            * ebosSimulator_.problem().porosity(c);
+                        fip_.fip[FIPData::FIP_PV][c] = pv;
```
We have calculated the pore volume three times now. I guess, each time we need to access two arrays now instead of one previously?

andlaus replied:
depends on what you mean with *calculated* since the `.model().dofTotalVolume(c)` and `problem().porosity(c)` methods are trivial. I'd say the pore volume was *accessed* three times here.

blattms commented on opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp:
```diff
@@ -157,24 +157,31 @@ class SimulatorFullyImplicitBlackoilEbos
         if( param_.getDefault("timestep.adaptive", true ) )
         {
-            adaptiveTimeStepping.reset( new AdaptiveTimeStepping( param_, terminal_output_ ) );
+
```
Hu? The FIP caculations are connected to the settings of the adaptive time stepper? That is quite a surprise. Please explain!

andlaus replied:
flow_legacy does it this way

blattms commented on opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp:
```diff
@@ -157,24 +157,31 @@ class SimulatorFullyImplicitBlackoilEbos
-            //                                    prev_well_state,
-            //                                    restorefilename,
-            //                                    desiredRestoreStep );
+            output_writer_.restore( timer,
```
and to this place, too?

andlaus replied:
flow_legacy again

blattms commented on opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp:
```diff
@@ -195,7 +202,6 @@ class SimulatorFullyImplicitBlackoilEbos
             }
         }
         std::vector<std::vector<double>> OOIP;
-
```
Even empty lines seem to influence it?!

andlaus replied:
no, but I suppose that the empty line is not there there in the flow_legacy version ;)

blattms commented on opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp:
```diff
@@ -221,47 +227,40 @@ class SimulatorFullyImplicitBlackoilEbos
             handleAdditionalWellInflow(timer, wells_manager, well_state, wells);
 
-            // Compute reservoir volumes for RESV controls.
-            computeRESV(timer.currentStepNum(), wells, state, well_state);
```
I am giving up...

andlaus commented:
> Please remove all changes that have nothing to do with #988.

not a big deal. as far as I can see, you mean the synchronization with FlowMainBase (which diverged a bit due to missing merges since FlowMainEbos was introduced)? I'll do as @atgeirr prefers.

andlaus commented:
ok, I've replaced all counting loops by loops over the grid. this has the advantages that the same mechanism to loop over all DOFs is used uniformly and that the parallel and non-parallel versions use the same code path. for Norne in sequential mode, the FIP numbers still match those of flow_legacy.

atgeirr commented:
> not a big deal. as far as I can see, you mean the synchronization with FlowMainBase (which diverged a bit due to missing merges since FlowMainEbos was introduced)? I'll do as @atgeirr prefers.

I think combining the "anti-diverging of FlowMain*" with "fix the FIP stuff" has lead to confusion. I think I want it all... So the best would be to split it in two PRs, then it all should be much clearer (and get merged more quickly). Agree?

andlaus commented:
agreed & PR split: #1026

atgeirr commented:
So are the FIP changes still in this PR, and ready for review?

andlaus commented:
> So are the FIP changes still in this PR, and ready for review?

yes, except for the change which is required to make #1026 not segfault anymore (and which leads to some rather simple merge conflicts with the present PR)

andlaus commented:
with #1026 merged, I've rebased this on top of the latest master to solve the merge conflicts. I suppose it is clean enough for review now...

andlaus commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1016,12 +1014,15 @@ namespace Opm {
                 const auto& fs = intQuants.fluidState();
+                const double pv =
+                    ebosSimulator_.model().dofTotalVolume(cellIdx)
+                    * ebosSimulator_.problem().porosity(cellIdx);
```
BTW: if the rock compressibility can be considered by simply using `intQuants.porosity().value()`.

atgeirr replied:
Starting over on this review, I may say things again that I or @blattms already have said. Bear over with me please...

Why not keep using geo_.poreVolume()?

atgeirr replied:
Also: did we have a clear answer to the question of including variable pore volume (rock compressibility)? I assume including it here would make flow_ebos behave different from flow_legacy?

andlaus replied:
> Why not keep using geo_.poreVolume()?

because its quite easy to switch away from it, with `geo_` it is much more difficult to also include rock compressibility if this turns out to be the Right Thing (TM) to do, it makes the parallel code less explicit (IMO) because it needs use non-standard mechanisms, and it forces the legacy object to be kept around even if it is not used by anything else...

andlaus replied:
> Also: did we have a clear answer to the question of including variable pore volume (rock compressibility)? I assume including it here would make flow_ebos behave different from flow_legacy?

the current version of the PR disregards rock compressibility to get a match with flow_legacy. Whether it should be included or not, I cannot really say...

atgeirr replied:
While I am perhaps clinging a bit close to the Eclipse-like way of thinking here, I think of (or have over time become conditioned to it) pore volume as consisting of a reference pore volume and a dynamic multiplier (pressure dependence). And in computing the reference pore volume there are lots of nitty-gritty details, that are not reflected in the apparent formula used here (volume times porosity). If it is correct for this particular purpose, then it must be because either the volume or the porosity must include such things as NTG. I assume it is baked into the porosity. In that case it will be wrong for various polymer simulations for example, which include an adsorption term that depends on porosity.

To sum up, I really don't like the explicit use of a formula here to compute pore volume, where before we used an already-digested number.

andlaus replied:
> And in computing the reference pore volume there are lots of nitty-gritty details, that are not reflected in the apparent formula used here (volume times porosity)

these nitty-gritty (NTG, MULTPV, MULTREGP, PORV, etc.) details are actually considered by the way the porosity is computed by EclProblem: https://github.com/OPM/ewoms/blob/master/ebos/eclproblem.hh#L982

As noted in OPM/ewoms#130, this is not very transparent and depends on lot of side conditions (mainly the ECL specification is to blame here, IMO) and on some details of the grid (e.g. how it implements MINPV).

>  I assume it is baked into the porosity.

exactly. I'm quick to admit that this is not exactly pretty, though.

> To sum up, I really don't like the explicit use of a formula here to compute pore volume, where before we used an already-digested number.

the reason why ebos uses porosity instead of pore volume is that the models in eWoms are not specific to the finite volume discretization and concept of a "pore volume" does not make any sense for discretizations that require the problem to be defined in terms of intensive quantities (e.g., finite element methods).

blattms commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1032,150 +1033,98 @@ namespace Opm {
-                // Compute the global dims value and resize values accordingly.
-                dims = comm.max(dims);
-                values.resize(dims, std::vector<double>(FIPData::fipValues,0.0));
+            comm.sum(hcpv.data(), hcpv.size());
```
Seems like these values have not even been initialized yet due to the changed order of the computations. That happens down in line 1088 ff.

atgeirr replied:
At this point, these vectors are just zeros, probably a bug.

andlaus replied:
yeah, quite possible. I did not think about whether the old version is semantically correct when refactoring it. ;)

andlaus replied:
right. hopefully it's fixed now...

atgeirr commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1032,150 +1033,98 @@ namespace Opm {
-                        }
-                    }
-                }
+            elemIt = elemCtx.gridView().template begin</*codim=*/0>();
```
I'll just refer to the comments I made before about this loop in a different location... Also, braces please (as in that other location...)

andlaus replied:
yeah, I was lost in PRs here. fixed, thanks.

atgeirr commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1032,150 +1033,98 @@ namespace Opm {
                 }
-#else
-                // This should never happen!
-                OPM_THROW(std::logic_error, "HAVE_MPI should be defined if we are running in parallel");
```
Why remove this?

andlaus replied:
it not necessary (the communicator is always used) and we would have to go through hoops to keep it (which is a weird thing to do for code which is never going to be executed IMO).

atgeirr commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1032,150 +1033,98 @@ namespace Opm {
-                            fip_.fip[FIPData::FIP_WEIGHTED_PRESSURE][c] = pres[region] / pv[c];
-                        }
+                elemCtx.updatePrimaryStencil(elem);
+                elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
```
Is this call the only reason why you change from the old-style loop with `c` to the new with (essentially) `elemCtx`?

andlaus replied:
the reason is primary that the "new" style works unconditionally, i.e. it works regardless of whether the cached intensive quantities are present or not (or in fact regardless of whether the intensive quantities are cached in the first place)

atgeirr commented:
Seems that @blattms and I both looked at this in parallel, happily no deadlock ensued...

blattms commented:
It would really help a lot to put the parallel stuff into a different PR!

andlaus commented:
> It would really help a lot to put the parallel stuff into a different PR!

you mean removing the calls to `comm`? that's easy but IMO it does not really make sense...

atgeirr commented:
OK, it is now time to resurrect this work and restart the discussion on it. I am not sure if it has been affected significantly by recent development, but I'd ask you to rebase it in any case to make it green again. Sorry for letting it lie still for so long...

andlaus commented:
ok, I've rebased this on top of the latest master and fixed the conflicts. as before rebasing the FIP numbers for flow ebos are much closer to those of flow_legacy with this PR but they are not quite a perfect match...

andlaus commented:
I've found a small bug in the calculation of the initial fluid in place. now the numbers are *very* close to those produced by ECL...

andlaus commented:
jenkins run norne_ebos please

atgeirr commented:
> review is stale and most issues have been addressed.

In general I think the PR author should not dismiss reviews... Although you are correct that its primary concerns have been addressed.

atgeirr commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1029,12 +1027,15 @@ namespace Opm {
                 const auto& fs = intQuants.fluidState();
+                const double pv =
+                    ebosSimulator_.model().dofTotalVolume(cellIdx)
+                    * intQuants.porosity().value();
```
Please add comment explaining why this is correct. Important points:
 - the intQuants.porosity() can be outside [0,1] due to pore volume manipulation
 - while the problem.porosity() would also account for this manipulation, only the intQuants version includes the pressure-dependent porosity multiplier

andlaus replied:
comment added. IMO the second point is self-evident because problem.porosity() returns a scalar instead of an evaluation. (i.e., it cannot be solution dependent.)

andlaus commented:
> In general I think the PR author should not dismiss reviews... Although you are correct that its primary concerns have been addressed.

yeah, I know and usually I don't. my main reason to do this was that it seems like @blattms dropped out of that discussion about three weeks ago.

atgeirr commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1045,150 +1046,100 @@ namespace Opm {
-                            values[region][phase] += fip_.fip[phase][c];
-                        }
-                    }
+            elemIt = elemCtx.gridView().template begin</*codim=*/0>();
```
Please keep the loop iterators in local scope.

andlaus replied:
do you mean to *create* a local scope around each of these loops to achive this?

andlaus replied:
ok: I've changed this so that the element iterators are created in the first part of the `for` statements. the result is harder to read IMO, so I would not mind if you requested to revert that patch before the PR gets merged.

atgeirr commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1045,150 +1046,100 @@ namespace Opm {
+                if (region != -1) {
+                    const auto& intQuants = elemCtx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
+                    const auto& fs = intQuants.fluidState();
+                    const double pv =
```
This formula is repeated very often, did you consider storing the value?

andlaus replied:
it is just a single multiplication, so storing it somewhere is not worth the hassle IMO. (also, it would slightly increase memory consumption and it may even be slightly slower due to cache effects; Yeah, we don't need to care about memory consumption and this part of the code is not very performance critical...)

atgeirr commented:
About the scoping: I think it's more readable now... Beauty is in the eye of the beholder! Intend to merge when Travis is green (don't think we need to rerun Jenkins).

andlaus commented:
> About the scoping: I think it's more readable now... Beauty is in the eye of the beholder!

ok, that's fine with me. (I don't care too much about it.)

> (don't think we need to rerun Jenkins).

I don't think so either; as far as I can see, nothing has changed functionality-wise since the last jenkins run.

andlaus commented:
that PR was a long story. thanks for merging :)
